Utilities/rANS/test/test_ransEncodeDecode.cxx

```diff
@@ -0,0 +1,96 @@
+  const uint32_t probabilityScale = 1 << P;
+
+  const std::string source =
+    R"(Sed ut perspiciatis, unde omnis iste natus error sit voluptatem accusantium 
+doloremque laudantium, totam rem aperiam eaque ipsa, quae ab illo inventore veritatis 
+et quasi architecto beatae vitae dicta sunt, explicabo. nemo enim ipsam voluptatem, 
+quia voluptas sit, aspernatur aut odit aut fugit, sed quia consequuntur magni dolores 
+eos, qui ratione voluptatem sequi nesciunt, neque porro quisquam est, qui dolorem ipsum, 
+quia dolor sit, amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora 
+incidunt, ut labore et dolore magnam aliquam quaerat voluptatem. ut enim ad minima veniam, 
+quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea 
+commodi consequatur? quis autem vel eum iure reprehenderit, qui in ea voluptate velit 
+esse, quam nihil molestiae consequatur, vel illum, qui dolorem eum fugiat, 
+quo voluptas nulla pariatur?)";
+};
+
+typedef boost::mpl::vector<Fixture<uint32_t, uint8_t, 14>, Fixture<uint64_t, uint32_t, 18>> Fixtures;
+
+BOOST_FIXTURE_TEST_CASE_TEMPLATE(test_EncodeDecode, T, Fixtures, T)
+{
+  // iterate over the message and create PDF and CDF for each symbol in the messsage
+  o2::rans::SymbolStatistics stats{std::begin(T::source), std::end(T::source)};
+  // For performance reason the distributions must be rescaled to be a power of two.
+  stats.rescaleFrequencyTable(T::probabilityScale);
+
+  // buffer to write the rANS coded message into
+  std::vector<typename T::stream_t> encoderBuffer(1 << 20, 0);
+  //create a stateful encoder object that builds an encoding table from the given symbol statistics
+  o2::rans::Encoder<typename T::coder_t> encoder{stats, T::probabilityBits};
+  // encoder rANS encodes symbols from SOURCE array to encoder Buffer. Since coder and decoder
+  // are mathematical inverse functions on the ANS state, they operate as a stack -
+  // i.e. the decoder outputs the message in reverse order of the encoder.
+  // By convention the encoder runs backwards over the input so that the decoder can return
+  // the data in the expected order. The encoded message ranges from ransBegin to encoderBuffer.end()
+  auto ransBegin = encoder.process(encoderBuffer.begin(), encoderBuffer.end(), std::begin(T::source), std::end(T::source));
+
+  // The decoded message will go into the decoder buffer which will have as many symbols as the original message
+  std::vector<typename T::source_t> decoderBuffer(std::end(T::source) - std::begin(T::source), 0);
+  // create a stateful decoder object that build a decoding table from the given symbol statistics
+  o2::rans::Decoder<typename T::coder_t, typename T::source_t> decoder(stats, T::probabilityBits, T::probabilityScale);
+  // the deocder unwinds the rANS state in the encoder buffer starting at ransBegin and decodes it into the decode buffer;
+  decoder.process(decoderBuffer.begin(), ransBegin, stats.getMessageLength());
```

Comment:
OK, thanks! I would be more clear if you would start decoding explicitly from the encoded message rather than from intermediate iterator.

Reply:
Thanks for the comment. The problem is that I can't know how large the encoded message will come out of the encoder and therefore I have to allocate a buffer that is "big enough".  Since the rANS encoder works backwards it will iterate backwards over the buffer until all symbols were encoded the position where it stops is `ransBegin` - so the actual encoded message goes from `ransBegin`to `encoderBuffer.end()`. It takes another copy to an adequatly sized buffer to pass this in a nicer way. I will document this better and improve the naming to be more obvious.